mmedit/models/inpaintors/aot_inpaintor.py

```diff
@@ -0,0 +1,252 @@
+            output['fake_img'] = fake_img
+
+        output['meta'] = None if 'meta' not in kwargs else kwargs['meta'][0]
+
+        if save_image:
+            assert save_image and save_path is not None, (
+                'Save path should been given')
+            assert output['meta'] is not None, (
+                'Meta information should be given to save image.')
+
+            tmp_filename = output['meta']['gt_img_path']
+            filestem = Path(tmp_filename).stem
+            if iteration is not None:
+                filename = f'{filestem}_{iteration}.png'
+            else:
+                filename = f'{filestem}.png'
+            mmcv.mkdir_or_exist(save_path)
+            img_list = [kwargs['gt_img']] if 'gt_img' in kwargs else []
+            img_list.extend(
+                [masked_img,
+                 mask.expand_as(masked_img), fake_res, fake_img])
+            img = torch.cat(img_list, dim=3).cpu()
+            self.save_visualization(img, osp.join(save_path, filename))
+            output['save_img_path'] = osp.abspath(
+                osp.join(save_path, filename))
+
+        return output
+
+    def train_step(self, data_batch, optimizer):
+        """Train step function.
+
+        In this function, the inpaintor will finish the train step following
+        the pipeline:
+        1. get fake res/image
+        2. compute reconstruction losses for generator
+        3. compute adversarial loss for discriminator
+        4. optimize generator
+        5. optimize discriminator
+
+        Args:
+            data_batch (torch.Tensor): Batch of data as input.
+            optimizer (dict[torch.optim.Optimizer]): Dict with optimizers for
+                generator and discriminator (if have).
+        Returns:
+            dict: Dict with loss, information for logger, the number of \
+                samples and results for visualization.
+        """
+        log_vars = {}
+
+        gt_img = data_batch['gt_img']
+        mask = data_batch['mask']
+        masked_img = data_batch['masked_img']
+        masked_img = masked_img.float() + mask
+
+        input_x = torch.cat([masked_img, mask], dim=1)
+        fake_res = self.generator(input_x)
+        fake_img = gt_img * (1. - mask) + fake_res * mask
+
+        # reconstruction losses
+        results, g_losses = self.generator_loss(fake_res, fake_img, data_batch)
+        loss_g, log_vars_g = self.parse_losses(g_losses)
+        log_vars.update(log_vars_g)
+
+        # adversarial loss
+        set_requires_grad(self.disc, True)
+        fake_data = fake_img.detach()
+        real_data = gt_img
+        disc_losses_real = self.forward_train_d(
+            real_data, True, True, mask=mask)
+        disc_losses_fake = self.forward_train_d(
+            fake_data, False, True, mask=mask)
+        disc_losses_ = disc_losses_real['real_loss'] + disc_losses_fake[
+            'fake_loss']
+        disc_losses = dict(disc_losses=disc_losses_)
```

Comment:
We usually separate the real_loss and fake_loss. You can refer to [here](https://github.com/open-mmlab/mmediting/blob/037efba0c80e40cb3a0f94b6a17e4621412db520/mmedit/models/inpaintors/one_stage.py#L363) for the implementation.

Reply:
the official implementation combined the real and fake losses together and then back propagate the total disc loss. Wonder if this would affect the results if we split the real and fake loss and back propagate each individually.

Reply:
I think it won't make a difference. But in any case we can follow the official implementation.